presto-main/src/main/java/com/facebook/presto/sql/rewrite/ShowStatsRewrite.java

```diff
@@ -135,52 +141,48 @@ protected Node visitNode(Node node, Void context)
-                    aliased(new Values(ImmutableList.of(new NullLiteral())),
-                            "table_stats_for_" + tableName,
-                            ImmutableList.of("column_name")));
-        }
-
-        private Constraint<ColumnHandle> getConstraint(TableScanNode node, Expression predicate, Map<Symbol, Type> types)
-        {
-            DomainTranslator.ExtractionResult decomposedPredicate = DomainTranslator.fromPredicate(
-                    metadata,
-                    session,
-                    predicate,
-                    types);
-
-            TupleDomain<ColumnHandle> simplifiedConstraint = decomposedPredicate.getTupleDomain()
-                    .transform(node.getAssignments()::get)
-                    .intersect(node.getCurrentConstraint());
-
-            return new Constraint<>(simplifiedConstraint, bindings -> true);
+            // analyze statement
+            QuerySpecification querySpecification = (QuerySpecification) node.getQuery().getQueryBody();
+            Query query = new Query(Optional.empty(), querySpecification, ImmutableList.of(), Optional.empty());
+            Analyzer analyzer = new Analyzer(session, metadata, sqlParser, accessControl, Optional.empty(), parameters);
+            Analysis analysis = analyzer.analyze(query);
+
+            // create domain
+            Expression filter = ((QuerySpecification) node.getQuery().getQueryBody()).getWhere().get();
+            Map<String, ColumnHandle> columnHandles = metadata.getColumnHandles(session, tableHandle);
+
+            // rewrite filter expression to use symbols in place of column references
+            final Map<QualifiedNameReference, Symbol> symbolMap = new HashMap<>();
+            Expression filterUsingSymbols = ExpressionTreeRewriter.rewriteWith(new ExpressionRewriter<Object>() {
+                @Override
+                public Expression rewriteQualifiedNameReference(QualifiedNameReference node, Object context, ExpressionTreeRewriter<Object> treeRewriter)
+                {
+                    String symbolName = node.getName().getSuffix();
+                    symbolMap.put(node, new Symbol(symbolName));
+                    return new SymbolReference(symbolName);
+                }
+            }, filter);
+
+            // compute types for symbols
+            Map<Symbol, Type> symbolTypes = symbolMap.entrySet().stream().collect(Collectors.toMap(
+                    Map.Entry::getValue,
+                    entry -> analysis.getType(entry.getKey())
+            ));
+
+            // compute tuple domain
+            TupleDomain<Symbol> symbolTupleDomain = DomainTranslator.fromPredicate(metadata, session, filterUsingSymbols, symbolTypes).getTupleDomain();
+
+            // transform computed tuple domain to use ColumnHandles
+            TupleDomain<ColumnHandle> columnTupleDomain = symbolTupleDomain.transform(symbol -> columnHandles.get(symbol.getName()));
```

Comment:
The column handles seem to be used only here. Move the declaration as close as possible to the usage.